worker/caasadmission/k8s/admission.go

```diff
@@ -0,0 +1,158 @@
+// Copyright 2020 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package k8s
+
+import (
+	"bytes"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+
+	"github.com/juju/errors"
+	admission "k8s.io/api/admissionregistration/v1beta1"
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/juju/juju/caas/kubernetes/provider"
+)
+
+type CertificateBroker interface {
+	// Register a new domain name lease on the brokers certificate. Returns a
+	// lease function if no errors
+	RegisterDomainLeases(...string) (func(), error)
+
+	// Set of root certificates in pem format
+	RootCertificates() ([]byte, error)
+}
+
+// TODO this is a tmp struct and won't make it to final
+type CertWatcherBroker struct {
+	CertWatcher func() *tls.Certificate
+}
+
+type K8sBroker interface {
+	EnsureMutatingWebhookConfiguration(*admission.MutatingWebhookConfiguration) (func(), error)
+}
+
+// Represents a creator of mutating webhooks that is context aware of the
+// current controller.
+type AdmissionCreator interface {
+	EnsureMutatingWebhookConfiguration() (func(), error)
+}
+
+// Func type of AdmissionCreator
+type AdmissionCreatorFunc func() (func(), error)
+
+const (
+	Component = "admission"
+)
+
+var (
+	anyMatch = []string{"*"}
+)
+
+// Implements AdmissionCreator interface for func type
+func (a AdmissionCreatorFunc) EnsureMutatingWebhookConfiguration() (func(), error) {
+	return a()
+}
+
+// Instantiates a new AdmissionCreator for the supplied context arguments.
+func NewAdmissionCreator(
+	certBroker CertificateBroker,
+	namespace, modelName, modelUUID string,
+	ensureConfig func(*admission.MutatingWebhookConfiguration) (func(), error),
+	service *admission.ServiceReference) (AdmissionCreator, error) {
+
+	caPems, err := certBroker.RootCertificates()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	failurePolicy := admission.Ignore
+	matchPolicy := admission.Equivalent
+	ruleScope := admission.AllScopes
+
+	// MutatingWebjook Obj
+	obj := admission.MutatingWebhookConfiguration{
```

Comment:
The other caas workers have managed to avoid using the k8s struct types directly - they call into the k8s broker interface passing in juju types and the k8s stuff is hidden under the covers. Same with watchers etc - the k8s types are mapped to juju types on the way out.

We could do the same here but I guess the controller and handler need to use the k8s types directly so there's little value in updating this one place.

Reply:
This is k8s specific so not much point. In fact trying to hide it would cause more issues as the design pattern of the k8s broker wouldn't work nicely for this.